### PR TITLE
Fix error handling loop

### LINKED ISSUE #10199: Exception loop in Engine during handling of unexpected error
**Describe the bug**

In #9946 we encountered the problem of an exception loop when we deployed a process that was too large. The cause for this that the processor did not handle this exception explicitly using the `tryHandleError` method. As a result the engine considered this an `UNEXPECTED_ERROR`. When this happens the engine will try to handle this unexpected error:
```java
  private void handleUnexpectedError(
      final Throwable processingException, final TypedRecord record) {
    final String errorMessage =
        String.format(
            ERROR_MESSAGE_PROCESSING_EXCEPTION_OCCURRED, record, processingException.getMessage());
    LOG.error(errorMessage, processingException);

    writers.rejection().appendRejection(record, RejectionType.PROCESSING_ERROR, errorMessage);
    writers
        .response()
        .writeRejectionOnCommand(record, RejectionType.PROCESSING_ERROR, errorMessage);
    errorRecord.initErrorRecord(processingException, record.getPosition());

    if (DbBlackListState.shouldBeBlacklisted(record.getIntent())) {
      if (record.getValue() instanceof ProcessInstanceRelated) {
        final long processInstanceKey =
            ((ProcessInstanceRelated) record.getValue()).getProcessInstanceKey();
        errorRecord.setProcessInstanceKey(processInstanceKey);
      }

      writers.state().appendFollowUpEvent(record.getKey(), ErrorIntent.CREATED, errorRecord);
    }
  }
```
At the beginning of this method we will create an error message and write the rejection. During the appending of this rejection we check the size of this record (https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/api/records/RecordBatch.java#L58). If it's also too large we will throw an exception.
Since this exception is also considered to be an `UNEXPECTED_ERROR` the Engine tries to handle this unexpected error again. This will try to append the record again and in turn thrown another exception, starting the loop again.

This loop will cause the partition to reach an UNHEALTHY state.

**To Reproduce**

As long as #10193 is not merged:
1. Deploy a very large process
2. The exception loop should occur

If #10193 is merged:
1. Change the `DeploymentCreateProcessor` `tryHandleError` method to always return an `UNEXPECTED_ERROR`
2. Deploy a very large process
3. The exception loop should occur

**Expected behavior**

The exception loop should not occur when the Engine fails to append a rejection.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
...with size: 57419999 this would exceed the maximum batch size. [ currentBatchEntryCount: 0, currentBatchSize: 0]
	at io.camunda.zeebe.engine.api.records.RecordBatch.appendRecord(RecordBatch.java:67) ~[classes/:?]
	at io.camunda.zeebe.streamprocessor.DirectProcessingResultBuilder.appendRecordReturnEither(DirectProcessingResultBuilder.java:76) ~[classes/:?]
	at io.camunda.zeebe.engine.api.ProcessingResultBuilder.appendRecord(ProcessingResultBuilder.java:36) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedRejectionWriter.appendRejection(ResultBuilderBackedRejectionWriter.java:31) ~[classes/:?]
	at io.camunda.zeebe.engine.Engine.handleUnexpectedError(Engine.java:174) ~[classes/:?]
	at io.camunda.zeebe.engine.Engine.onProcessingError(Engine.java:161) ~[classes/:?]
	at io.camunda.zeebe.streamprocessor.ProcessingStateMachine.lambda$errorHandlingInTransaction$7(ProcessingStateMachine.java:336) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[classes/:?]
	at io.camunda.zeebe.streamprocessor.ProcessingStateMachine.errorHandlingInTransaction(ProcessingStateMachine.java:324) ~[classes/:?]
	at io.camunda.zeebe.streamprocessor.ProcessingStateMachine.lambda$onError$6(ProcessingStateMachine.java:313) ~[classes/:?]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) ~[classes/:?]
```

</p>
</details>

**Environment:**
- OS: Cloud
- Zeebe Version: Starting from 8.1.0-alpha4

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

When we exceed the record batch an exception is thrown. When this is done during the handling of an error this would result in the writing of a rejection to fail, in turn resulting in an exception-loop.

By removing the rejection reason from the rejection in the event of an `ExceededBatchRecordSizeException` the rejection should always be able to be written. Unfortunately removing this reason makes it unclear what went wrong. To still be able to identify this a new rejection type has been added.

## Related issues

closes #10199 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/Engine.java [hunk 1]
```diff
@@ -171,10 +172,15 @@ private void handleUnexpectedError(
             ERROR_MESSAGE_PROCESSING_EXCEPTION_OCCURRED, record, processingException.getMessage());
     LOG.error(errorMessage, processingException);
 
-    writers.rejection().appendRejection(record, RejectionType.PROCESSING_ERROR, errorMessage);
-    writers
-        .response()
-        .writeRejectionOnCommand(record, RejectionType.PROCESSING_ERROR, errorMessage);
+    if (processingException instanceof ExceededBatchRecordSizeException) {
+      writers.rejection().appendRejection(record, RejectionType.EXCEEDED_BATCH, "");
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/Engine.java:176]: :wrench: We could add a comment about why the rejection message is empty.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceRejectionTest.java [hunk 2]
```diff
@@ -20,12 +20,14 @@
...
 import io.camunda.zeebe.test.util.record.RecordingExporter;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
+import io.camunda.zeebe.util.ByteValue;
 import org.assertj.core.api.Assertions;
 import org.junit.Rule;
 import org.junit.Test;
 
 public class CreateProcessInstanceRejectionTest {
 
+  public static final long MAX_MESSAGE_SIZE = ByteValue.ofMegabytes(4);
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceRejectionTest.java:30]: :wrench: We could change the visibility of the constant.

```suggestion
  private static final long MAX_MESSAGE_SIZE = ByteValue.ofMegabytes(4);
```

File: protocol/src/main/resources/protocol.xml [hunk 3]
```diff
@@ -60,6 +60,7 @@
       <validValue name="ALREADY_EXISTS">2</validValue>
       <validValue name="INVALID_STATE">3</validValue>
       <validValue name="PROCESSING_ERROR">4</validValue>
+      <validValue name="EXCEEDED_BATCH">5</validValue>
```
Comment (saig0) [inline protocol/src/main/resources/protocol.xml:63]: :wrench: We could change the name of the rejection type to `EXCEEDED_BATCH_RECORD_SIZE`.

Currently, we print only the rejection type in the client exception. If the rejection type is more self-descriptive then the exception is easier to understand.

:thinking: In the configuration, the size is called max. message size. But I think the term `message` could be confusing. The underlying protocol is based on messages but we try to write a batch of records.

**Review by saig0 — APPROVED** [review]
@remcowesterhoud looks nice. :+1: 

I have a few minor suggestions. Please have a look.

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors merge

**Pull request merged**